File: DP/LIS/boj_가장 긴 증가하는 부분 수열.cpp
//www.acmicpc.net/problem/11053
#include<iostream>
#include<algorithm>
using namespace std;

int arr[1001];
int dp[1001];
int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);


    int n;
    cin >> n;
    for (int i = 1; i <= n; i++) {
        cin >> arr[i];
    }
    dp[1] = 1;
    int ans = 1;
    for (int i = 2; i <= n; i++) {
        for (int j = i - 1; j >= 0; j--) {
            if (arr[i] > arr[j]) {
                if (dp[i] <= dp[j]) {
                    dp[i] = dp[j] + 1;

                    ans = max(ans, dp[i]);
                }

            }
        }
    }
    cout << ans;

    return 0;
}

File: DP/LIS/boj_가장 긴 증가하는 부분 수열.kt
//www.acmicpc.net/problem/11053
import java.util.*
import kotlin.math.*
fun main() = with(System.out.bufferedWriter()){
    val br = System.`in`.bufferedReader()
    val n = Integer.parseInt(br.readLine())
    val arr = IntArray(n)
    val dp = IntArray(n)
    val tk = StringTokenizer(br.readLine())
    dp[0]=1
    arr[0] = Integer.parseInt(tk.nextToken())
    var answer=1
    for(i in 1 until n){
        arr[i] = Integer.parseInt(tk.nextToken())
        for(j in i-1 downTo 0){
            if(arr[i]>arr[j]){
                dp[i]= max(dp[i],dp[j]+1)
            }
        }
        if(dp[i]==0){
            dp[i]=1
        }
        answer = max(answer,dp[i])
    }
    write("$answer")
    close()
}
